File: cli/backend/meteo/meteo_forecast.py
import json
import math

import core

from cli import WeatherForecast
from cli.backend.meteo.meteo_current import MeteoCurrent
from cli.backend.meteo.meteo_future import MeteoFuture


class MeteoForecast(WeatherForecast):
    def __init__(self, loc, metric):
        location = self.get_location(loc)
        country = location.raw["address"]["country"]
        region = location.raw["address"]["city"]
        if not metric:
            data = core.networking.get_urls(
                [
                    "https://api.open-meteo.com/v1/forecast?latitude={}&longitude={}&current_weather=true"
                    "&hourly=temperature_2m,rain,showers,snowfall,cloudcover,dewpoint_2m,apparent_temperature,"
                    "pressure_msl,visibility,windspeed_10m,winddirection_10m"
                    "&daily=temperature_2m_max,temperature_2m_min"
                    "&temperature_unit=fahrenheit&windspeed_unit=mph&precipitation_unit=inch"
                    "&timezone=auto".format(loc[0], loc[1]),
                    "https://air-quality-api.open-meteo.com/v1/air-quality?latitude={}&longitude={}&hourly=european_aqi"
                    "".format(loc[0], loc[1]),
                ]
            )
        else:
            data = core.networking.get_urls(
                [
                    "https://api.open-meteo.com/v1/forecast?latitude={}&longitude={}&current_weather=true"
                    "&hourly=temperature_2m,rain,showers,snowfall,cloudcover,dewpoint_2m,apparent_temperature,visibility"
                    ",windspeed_10m,winddirection_10m"
                    "&daily=temperature_2m_max,temperature_2m_min&timezone=auto".format(
                        loc[0], loc[1]
                    ),
                    "https://air-quality-api.open-meteo.com/v1/air-quality?latitude={}&longitude={}&hourly=european_aqi"
                    "".format(loc[0], loc[1]),
                ]
            )
        forecast_json = json.loads(data[0])
        aqi_json = json.loads(data[1])
        raw_data = [forecast_json, aqi_json]
        forecast = [MeteoCurrent(forecast_json, aqi_json, metric)]
        for i in range(forecast[0].now + 1, len(forecast_json["hourly"]["rain"])):
            forecast.append(MeteoFuture(forecast_json, aqi_json, metric, i))
        super().__init__(0, region, country, forecast, "", raw_data)
        self.forecast_sentence = self.get_forecast_sentence()

    def get_forecast_sentence(self):
        rain = [amount != 0 for amount in self.raw_data[0]["hourly"]["rain"]]
        snow = [amount != 0 for amount in self.raw_data[0]["hourly"]["snowfall"]]
        for i in range(self.forecast[0].now):
            rain.pop(0)
            snow.pop(0)
        if True in [
            condition.condition_id // 100 == 5
            for condition in self.forecast[0].conditions
        ]:
            t = 0
            for i in rain:
                if not i:
                    break
                t += 1
            return "It will continue raining for " + str(t) + " hours."
        if True in [
            condition.condition_id // 100 == 6
            for condition in self.forecast[0].conditions
        ]:
            t = 0
            for i in snow:
                if not i:
                    break
                t += 1
            return "It will continue snowing for " + str(t) + " hours."
        else:
            if True in rain:
                rain_start = rain.index(True)
            else:
                rain_start = math.inf
            if True in snow:
                snow_start = snow.index(True)
            else:
                snow_start = math.inf
            if rain_start == math.inf and snow_start == math.inf:
                return "Conditions are predicted to be clear for the next 7 days."
            rain.reverse()
            snow.reverse()
            if rain_start != math.inf:
                rain_end = rain.index(True)
            else:
                rain_end = math.inf
            if snow_start != math.inf:
                snow_end = snow.index(True)
            else:
                snow_end = math.inf
            if rain_start != math.inf:
                return (
                    "It will rain in "
                    + str(rain_start)
                    + " hours for "
                    + str(rain_end - rain_start)
                    + " hours"
                )
            if snow_start != math.inf:
                return (
                    "It will rain in "
                    + str(snow_start)
                    + " hours for "
                    + str(snow_end - snow_start)
                    + " hours"
                )
        return "Conditions are predicted to be clear for the next 7 days."

File: src/backend/meteo/meteo_forecast.rs
use crate::backend::meteo::meteo_current::get_meteo_weather_data;
use crate::backend::meteo::meteo_get_combined_data_formatted;
use crate::backend::meteo::meteo_json::MeteoForecastJson;
use crate::backend::weather_data::WeatherDataRS;
use crate::location;
use crate::backend::weather_forecast::WeatherForecastRS;
use crate::local::settings::Settings;

fn get_forecast_sentence(
    data: Vec<WeatherDataRS>,
    raw_data: MeteoForecastJson,
    start: usize,
) -> String {
    let mut rain = raw_data
        .hourly
        .rain
        .iter()
        .map(|x| (x - 0.0).abs() > f32::EPSILON)
        .collect::<Vec<bool>>();
    let mut snow = raw_data
        .hourly
        .snowfall
        .iter()
        .map(|x| (x - 0.0).abs() > f32::EPSILON)
        .collect::<Vec<bool>>();
    rain.drain(0..start);
    snow.drain(0..start);
    if data[0]
        .conditions
        .clone()
        .into_iter()
        .map(|condition| condition.condition_id / 100 == 5)
        .collect::<Vec<bool>>()
        .contains(&true)
    {
        let mut t: u8 = 0;
        for i in rain {
            if !i {
                break;
            }
            t += 1;
        }
        return format!("It will continue raining for {t} hours.");
    }
    if data[0]
        .conditions
        .clone()
        .into_iter()
        .map(|condition| condition.condition_id / 100 == 6)
        .collect::<Vec<bool>>()
        .contains(&true)
    {
        let t = snow.iter().position(|&b| b).unwrap_or(0);
        return format!("It will continue snowing for {t} hours.");
    }
    let rain_start = rain.clone().into_iter().position(|x| x);
    let snow_start = snow.clone().into_iter().position(|x| x);

    if rain_start.is_none() && snow_start.is_none() {
        return "Conditions are predicted to be clear for the next 7 days.".to_string();
    }
    rain.reverse();
    snow.reverse();
    let rain_end = rain.into_iter().position(|x| x);
    let snow_end = snow.into_iter().position(|x| x);
    if rain_start.is_some() {
        return format!(
            "It will rain in {} hours for {} hours",
            rain_start.unwrap(),
            rain_end.unwrap() - rain_start.unwrap()
        );
    }
    if snow_start.is_some() {
        return format!(
            "It will snow in {} hours for {} hours",
            snow_start.unwrap(),
            snow_end.unwrap() - snow_start.unwrap()
        );
    }
    String::from("Conditions are predicted to be clear for the next 7 days.")
}

pub fn get_meteo_forecast(
    coordinates: [&str; 2],
    settings: Settings,
) -> crate::Result<WeatherForecastRS> {
    let data = meteo_get_combined_data_formatted(coordinates, settings.internal.metric_default)?;
    let mut forecast: Vec<WeatherDataRS> = Vec::new();
    let now = data
        .weather
        .hourly
        .time
        .iter()
        .position(|r| *r == data.weather.current_weather.time)
        .expect("now not found");
    let current = get_meteo_weather_data(
        data.weather.clone(),
        data.air_quality.clone(),
        now,
        settings.internal.metric_default,
    )?;
    forecast.push(current);
    for i in now + 1..data.weather.hourly.time.len() - 1 {
        forecast.push(get_meteo_weather_data(
            data.weather.clone(),
            data.air_quality.clone(),
            i,
            settings.internal.metric_default,
        )?);
    }
    let region_country = location::reverse_geocode(coordinates[0], coordinates[1])?;
    let forecast_sentence = get_forecast_sentence(forecast.clone(), data.weather, now);
    let f = WeatherForecastRS {
        region: region_country[0].clone(),
        country: region_country[1].clone(),
        forecast: forecast.clone(),
        current_weather: forecast.into_iter().next().unwrap(),
        forecast_sentence,
        raw_data: None,
    };
    Ok(f)
}
